src/analytics-pre-init.ts

```diff
@@ -0,0 +1,307 @@
+import { Analytics } from './analytics'
+import { Context } from './core/context'
+import { allSettled } from './lib/all-settled'
+
+/**
+ * The names of any Analytics instance methods that can be called pre-initialization.
+ * These methods should exist statically on AnalyticsBrowser.
+ */
+type PreInitMethodName =
+  | 'trackSubmit'
+  | 'trackClick'
+  | 'trackLink'
+  | 'trackForm'
+  | 'pageview'
+  | 'identify'
+  | 'reset'
+  | 'group'
+  | 'track'
+  | 'ready'
+  | 'alias'
+  | 'debug'
+  | 'page'
+  | 'once'
+  | 'off'
+  | 'on'
+  | 'addSourceMiddleware'
+  | 'addIntegrationMiddleware'
+  | 'setAnonymousId'
+  | 'addDestinationMiddleware'
+
+export function flushAnalyticsCallsInNewTask(
+  analytics: Analytics,
+  buffer: PreInitMethodCallBuffer
+): void {
+  buffer.toArray().forEach((m) => {
+    setTimeout(() => {
+      callAnalyticsMethod(analytics, m).catch(console.error)
+    }, 0)
+  })
+}
+
+async function flushAnalyticsCallsConcurrentlyByName(
+  name: PreInitMethodName,
+  analytics: Analytics,
+  buffer: PreInitMethodCallBuffer
+): Promise<void> {
+  await allSettled(
+    buffer.getCalls(name).map((c) => {
+      return callAnalyticsMethod(analytics, c).catch(console.error)
+    })
+  )
+}
+
+async function flushAnalyticsCallsSeriallyByName(
+  name: PreInitMethodName,
+  analytics: Analytics,
+  buffer: PreInitMethodCallBuffer
+): Promise<void> {
+  for (const c of buffer.getCalls(name)) {
+    await callAnalyticsMethod(analytics, c).catch(console.error)
+  }
+}
+
+export const flushAddSourceMiddleware = flushAnalyticsCallsSeriallyByName.bind(
+  this,
+  'addSourceMiddleware'
+)
+
+export const flushOn = flushAnalyticsCallsConcurrentlyByName.bind(this, 'on')
+
+export const flushSetAnonymousID = flushAnalyticsCallsConcurrentlyByName.bind(
+  this,
+  'setAnonymousId'
+)
```

Comment:
Yep, this doesn't change the previous behavior in terms of how these are flushed.

The previous implementation flushed the `on` methods in basically the same way, the fact of its concurrency was just obscured by the implementation -- since it was in a forEach which doesn't "await" between iterations. You can refactor any forEach(async () =>) to a `promise.allSettled` without changing the iteration behavior, and you get the added benefit of the option of awaiting the entire operation when you're done (we're not doing this all the time, but it's an option). I'm just refactoring to be explicit here, since the previous implementation had a wide range of async behaviors / treatments.

setAnonymousID is a synchronous function if analytics is instantiated, but if a user calls it during pre-init, it returns a promise that will resolve when it gets flushed -- so it needs to be called in an async way.

Reply:
So about `on`, is this the method we are calling: https://github.com/segmentio/analytics-next/blob/master/src/core/emitter/index.ts#L4-L7 ? This is what confuses me, as it seems to be just adding the callback to the array synchronously 🤔

Reply:
Yeah! too many methods with different signatures 😓  Good opportunity to trim the runConcurrently utility, and probably remove the import of AllSettled to save on bundle size